### ISSUE TITLE
Indoor Navigation

### ISSUE THREAD

**Issue by dmenneck** [valhalla/valhalla#3650]
I am currently facing the use case of navigating in buildings. Unfortunately I couldn’t find any examples so far and wanted to ask if some of you have already made experience with it and/or can estimate if indoor navigation with Valhalla is possible.

The underlying OSM data will be tagged with schemas like IndoorOSM or Simple Indoor Tagging.

**Comment by kevinkreiser** [comment]
@dmenneck it is possible. in #3519 we added some basic support for this. i'm neither familiar with "IndoorOSM" nor "Simple Indoor Tagging" but if it uses the osm data model and the tags are ones that our processing cares about it should somewhat work. you can always give it a try and see how far it goes and what shortcomings you find. it would be great to cataloge a list of that here so we can address it as more of a larger project.

**Comment by dmenneck** [comment]
Thank you @kevinkreiser !

Is there any documentation somewhere where I can find the tags that Valhalla expects for Indoor Navigation? Which OSM tags should I use for mapping a building and making Valhalla able to navigate over the graph?

**Comment by nilsnolde** [comment]
You can look at the graph lua file in this PR which added the indoor stuff to the graph: https://github.com/valhalla/valhalla/pull/3509/files, which is fairly straight forward to derive from.

**Comment by dmenneck** [comment]
You‘re right @nilsnolde, it’s actually straight forward, thanks!

Since Valhalla can handle stairs/elevators/escalators it’s obviously possible to navigate from one floor to another. I’m curious how a request would be structured to navigate eg from the first floor to the fourth.

Do you have an example for that?

**Comment by kevinkreiser** [comment]
@dmenneck the unit tests in the pr linked above does some test requests but at the present time we dont have a way to select what floor you are on if you want to set a destination on a certain floor. we had some discussion about it before but its very similar to the "level" parameter we would use to select a road when multiple roads are on top of one another. if your question is focused on this problem the answer at the moment is we've not yet implemented the ability to select the "level" or "floor" but if the destination or origin that you pick is clearly on top of one walk way vs another it will pick that one regardless of the floor its on. for ambiguous cases where they are on top of each other it will depend on the ordering of the data unfortunately.

**Comment by kevinkreiser** [comment]
both work should work already but the problem is when hallways on two different floors are on top of each other, in that ambiguous case there is no way to tell the router which floor you meant

**Comment by kevinkreiser** [comment]
oh and we also dont support "area routing" which is that if there isnt a mapped walking path/hallway/stairway ie a linear feature then we cannot route on it.

**Comment by nilsnolde** [comment]
> would giving an offset to a hallway solve this issue?

jep, that'd do it. not really a universal solution for OSM but if you're assembling your own data then no problem. the way which is closest to the input coordinate will be selected for routing. so you really have to control the input to valhalla very tightly, a user can't just click on a room if you can't control which way it really snaps to, the room a user wants might on the third floor, but the closest way to the input point could be in the basement.

**Comment by kevinkreiser** [comment]
@pvormittag great that you are interested. it would be helpful if you could enumerate the issues that you've encountered thus far so we can discuss mitigations or features to address these and plan how the implementations should look in at least in broad strokes.

taking levels into account is already a bigger one that could at least touch: 

1. topology
2. edge candidate filtering
3. output so that you can do something visually

and we should try to tackle these one at a time and have a bit of an overview on how to do each

**Comment by nilsnolde** [comment]
I thought I remembered that topology was already accounting for elevators/escalators, but no input/output of levels so the only thing that’s actually working is the narrative & geometry (point 2 & 3)?

**Comment by kevinkreiser** [comment]
@nilsnolde thats right topology should be handled just by osms model but i didnt want to leave it off because i have the feeling that what we have is probably not in its final form even in that regard, but id love to be happily surprised :smile: 

@pvormittag yep thats the last two things i had in my list the second one to specify, for each location, what level you actually want and the second one setting something in the output so that the caller can properly visualize or do whatever application side logic that cares about which parts of the route are which levels.

for the former i think we can support asking for a specific level and modifying the candidate search to care via a filter. i suppose its a bit tricky since the way isnt on a level just the nodes. so we will have to decide if an edge is a candidate if at least one of its nodes is on the right level or something else. for that to even work we have to break the graph and osm nodes that have a level tag, im not sure if we are doing that in the parser right now but i dont remember doing that (and once we do do that that would be a topological change (inserting nodes in the graph at osm nodes that have the level tag)).

the latter is a big design question. im not sure id want to break up the maneuvers so that they match all levels traversed. you might get something like: take the ground floor elevator to the first floor, take the first floor elevator to the second floor, ..., take the nth floor elevator to the (n+1)th  floor. for that reason my gut feeling is that we should annotate the shape with (sparse) level information rather than chop up the route/directions based on the levels. you can think of level information as an overlay on the existing route/directions output structure. would something like that work for you? where you get information about what points in the shape are what levels?

eg. 1: a route leg that took place entirely at ground level and had 10 shape points total would look like: `[[0, 0]]`

eg. 2: a route leg that started on the ground floor, took the escalator to the 2nd floor (but had a node where you could get off at the 1st floor) and grabbed the elevator to the 4th floor might look like: `[[0, 0], [2, 1], [3, 2], [4, 3], [5, 4]]'

basically telling you when the level changes by telling you first the shape index it changes at followed by the level it changed to. i guess technically the second thing would have to be a string potentially but you get the gist.

this is just off the top of my head

**Comment by pvormittag** [comment]
@kevinkreiser,

> for the former i think we can support asking for a specific level and modifying the candidate search to care via a filter.

Yes, filtering for candidates given level criteria makes intuitive sense to me and also makes sense in the context of a search space. So I think we're aligned on this point.

> the latter is a big design question. I'm not sure id want to break up the maneuvers so that they match all levels traversed...

This is certainly the trickier part and the right direction for a good implementation is much less obvious. I'm unsure how maneuvers are exactly modeled, but I assume they represent a single step in a multi-step set of instructions. So in thinking from a person attempting to both understand the route visualization (e.g. the line shape drawn over the map context)  while also reading the step instructions (turn-by-turn), they need to align and make sense.

My first thought is for any node that connects to multiple levels -- regardless of their type and simply the fact they are mapped as a single OSM node -- would represent breakpoints, which I believe is @nilsnolde's insight of being "input/output of levels". However there is also the instruction to the user. So walking on a path segment to the elevator is a step, *taking the elevator is a step*, then walking from the elevator on the next path segment is a step. So this would only result in two path segments, but 3 steps to the user.

Even if the application is able to know which part(s) of the shape are on which level (and this does help and is still needed), it doesn't solve the situation, as we can't really visualize the user taking the elevator other than highlighting that node, which is what I think should happen for this case. The maneuver states "Take the elevator to the Level 2", but you really only want to highlight that single node.

[Considering the image in my comment above](https://github.com/valhalla/valhalla/issues/3650#issuecomment-2111109076), the maneuver narrative is "Take the elevator to the Level 2.", but the shape data the maneuver is aligned to doesn't make sense in that narrative context, which seems to be due to the elevator node always becoming the start node of the next path segment on the route.

---

If you have enough to take a shot at an implementation, I'm happy to build from a branch and integrate the changes into the demo above.

**Comment by kevinkreiser** [comment]
yep on the second part I think we first need to clear some way for nomenclature discussions. a maneuver in vakhalla parlance is a subsection of the route that can be described with an enumeration and metadata (maneuver type, street name etc) as well as some natural language instruction (eg take the elevator to the 5th floor). so yes i didnt mean that we should preclude level changes from maneuver generation cirteria. In fact i agree that SOME maneuvers should happen when there is a change in level. what I'm saying is that not ALL level changes have to be maneuvers and because of that we need to supply levels metadata (ie a 3rd dimension for the shape) so that applications can have precise level data as well as maneuvers that include this info optionaly.

if you model an elevator as a single way with nodes at each level. you don't want a maneuver happening at each level. you want one when you get in the elevator and one when you get out. this is just a conjured example but there will certainly be more plausible ones where we want to know what's going on with the levels with respect to shape but we don't need a maneuver each time.

**Comment by pvormittag** [comment]
Thanks for the clarification!

If the level data is provided as a sparse overlay on the route shape, I think this is fine as long as you can determine for every node in the shape which level its associated to. This allows the application to ensure it only renders the part of the shape for any given level.

Given maneuvers already provide their relationship to the shape (via `begin_shape_index` and `end_shape_index`), then this does seem like the route API would provide enough information to construct a useful visual and interesting interactions.

That said, I'm not sure I fully understand what this list of pairs represent then, especially if its defined at the route leg. If the second element is the level, what is the first element represent?

>  eg. 1: a route leg that took place entirely at ground level and had 10 shape points total would look like: `[[0, 0]]`

> eg. 2: a route leg that started on the ground floor, took the escalator to the 2nd floor (but had a node where you could get off at the 1st floor) and grabbed the elevator to the 4th floor might look like: `[[0, 0], [2, 1], [3, 2], [4, 3], [5, 4]]'

**Comment by kevinkreiser** [comment]
Ah sorry I didn't spell that out very well! The first element would be the shape index the level (second element) maps to. So that we have a "changelog" that tells us at which shape index the level changes to a new value.

**Comment by pvormittag** [comment]
Ah ok, makes sense.

For the proposed structure, I'll define out below how an application would need to expand the data to fully map the level to  each point in a route leg shape. I'll use "point" when referring a shape's 2D vector. Let me know if this aligns with your thinking!

> eg. 1: a route leg that took place entirely at ground level and had 10 shape points total would look like: [[0, 0]]

The form `[[0, n]]`, where `0` is the start index of the shape points array and `n` is a level value, is effectively a special case. Any time the list contains a single tuple, it indicates all the points for the shape are on the same level (as you mention). So for `[[0, 0]]` on a shape with 10 points, it expands to `[[0, 0], [1, 0], [2, 0], [3, 0], [4, 0], [5, 0], [6, 0], [7, 0], [8, 0], [9, 0]]`.

> eg. 2: a route leg that started on the ground floor, took the escalator to the 2nd floor (but had a node where you could get off at the 1st floor) and grabbed the elevator to the 4th floor might look like: `[[0, 0], [2, 1], [3, 2], [4, 3], [5, 4]]'

This example is for a shape with 6 points which traverses 5 levels, would expand to: `[[0, 0], [1, 0,], [2, 1], [3, 2], [4, 3], [5, 4]]`.

Just thinking this out in my head, I'm not sure if this expanded data is enough for rendering purposes. It seems like you'd also have to reify certain points (aka repeat some points) as part of rendering, in order to show segments of the shape on each level. If you only render what's expanded in the example, the line up the escalator would be fully drawn on the ground floor and nothing would be drawn on the first floor. This may be the key challenge to figuring this out, because with a route shape on a single 2D plane, there is no need to repeat points, but that doesn't seem to be the case for stacked 2D planes.

**Comment by chrstnbwnkl** [comment]
Regarding the location candidate search, it seems that the information we need is mostly in place already (namely, level information on the edges).
 
What's still missing is to account for edges that connect various levels (e.g. stairs/escalators) so that the location search can correctly identify edge candidates "between" levels. OSM's [Simple Indoor Tagging](https://wiki.openstreetmap.org/wiki/Simple_Indoor_Tagging) scheme allows for various semicolon separated entries for the `level` and `level:ref` values. Valhalla currently does not break the values up and just stores them as one string. 

Regarding the shape level change indices, I think it would be necessary to add level information to `NodeInfo`, as @dnesbitt61 pointed out. That structure is running out of space though, so there is no straight forward way to accomplish this... the only possibility I see would be to re-use the `elevation` bits for nodes that have `indoor=yes` and `level=*` tags, as elevation data is probably of little use anyway for indoor objects? But that will definitely mess with older Valhalla versions, so I'm eager to hear other suggestions :smile:
